### PR TITLE
FEATURE: Chat global mention warnings (pre-send & post-send)

### PR DESCRIPTION
**Pull request by markvanlan**
This adds a warning to the chat composer when `@all` or `@here` are added to the composer and the channel has global mentions disabled

Screenshot 2023-07-24 at 11 28 19 AM

Then after send, there is another mention warning:

Screenshot 2023-07-24 at 11 28 26 AM

___

This is also fixes the issue of chat composer warnings persisting across channels. Currently if you try to mention more groups than is allowed for example, a mention warning pops up. When you change channels the mention warning will not disappear even if there is no text in the composer.

This adds a `reset` function to the [chat-composer-warnings-tracker.js](https://github.com/discourse/discourse/compare/chat-global-mentions-warning?expand=1#diff-d1e8c87a149513a257202dad3b0f4770e46e549bf49b54c9f612fe404aaee042), which is called when the channel is changed and the message is empty. In the event that the message is not empty we call `captureMentions` to check the loaded drafts' mentions.

This PR would be nicer if the post-send notice used the new [chat notices](https://github.com/discourse/discourse/commit/3171fd1a0ac0256d772227149a2e73d2cc763788) API to publish the mention warnings but we would have to change the existing ones and I thought that would be too much change for this PR. It'd be a good followup though.

### CODE REVIEW

File: plugins/chat/assets/javascripts/discourse/services/chat-composer-warnings-tracker.js [hunk 1]
```diff
@@ -32,17 +33,33 @@ export default class ChatComposerWarningsTracker extends Service {
...
   @bind
-  trackMentions(message) {
+  reset() {
+    this.unreachableGroupMentions = [];
+    this.unreachableGroupMentions = [];
+    this.overMembersLimitGroupMentions = [];
+    this.tooManyMentions = false;
+    this.channelWideMentionDisallowed = false;
+    this.mentionsCount = 0;
+    this.mentionsTimer = null;
```
Comment (jjaffeux) [inline plugins/chat/assets/javascripts/discourse/services/chat-composer-warnings-tracker.js:43]: we should cancel instead I think here

**Comment by jjaffeux** [comment]
Cool, I never look too much into this code, I think we should change it, a service is probably not appropriate. It should be an object associated to the lifetime of the composer and partially recomputed on input change.

But totally fine with the addition and fix for now, thanks :+1:

File: plugins/chat/spec/system/mention_warnings_spec.rb [hunk 2]
```diff
@@ -69,5 +70,42 @@
...
+        )
+        chat_page.visit_channel(channel_1)
+
+        # Channel 1 has a draft that causes a mention warning. Should appear on load
+        expect(page).to have_css(".chat-mention-warnings")
+        expect(page.find(".chat-mention-warnings-list__simple")).to be_present
+
+        # Channel 2 doesn't have a draft so it should disappear
+        chat_page.visit_channel(channel_2)
+        expect(page).not_to have_css(".chat-mention-warnings")
```
Comment (jjaffeux) [inline plugins/chat/spec/system/mention_warnings_spec.rb:101]:
```suggestion
        expect(page).to have_no_css(".chat-mention-warnings")
```

Comment (jjaffeux) [reply plugins/chat/spec/system/mention_warnings_spec.rb:101]: This will be slow otherwise, as we will wait full duration to ensure it's not present

**Comment by jjaffeux** [comment]
@AndrewPrigorshnev could you review this one please?

**Review by AndrewPrigorshnev — COMMENTED** [review]
I left some comments, but it looks great generally 👍.

**Comment by jjaffeux** [comment]
@markvanlan did you see the comments from @AndrewPrigorshnev ?

**Comment by markvanlan** [comment]
@AndrewPrigorshnev Can you review again? I pushed a commit to address passing in the entire `currentMessage` rather than breaking up its attrs and passing in, as well as other little things.

The `skipDebouce` isn't a problem. If you log `captureMentions` calls it _looks_ like network requests are happening all the time but it's actually not. The skip debounce is passed in true or false and on first load it's important to skip the debounce when loading a draft so the composer isn't left with a mention warning for almost half a secod that doesn't apply. And when they start typing it's exactly the same perf as now.

**Comment by jjaffeux** [comment]
Ok great :+1:

> This PR would be nicer if the post-send notice used the new [chat notices](https://github.com/discourse/discourse/commit/3171fd1a0ac0256d772227149a2e73d2cc763788) API to publish the mention warnings but we would have to change the existing ones and I thought that would be too much change for this PR. It'd be a good followup though

@markvanlan feel free to take this if you want to

**Pull request merged**